### PR TITLE
Add Teams model 

### PR DESCRIPTION
**Pull request by Rashmi-K-A**
This PR is in reference to issue #548 for creating a ~Groups~ **Teams** model to handle organizational hierarchies. This PR includes changes to the following files:

- models.py - add Team model
- db.py - add internal methods `add_team`, `find_team` and `delete_team`
- api.py - add methods `add_team`, `delete_team`

#541  has more information on the discussion around how to handle organizational hierarchies.

### CODE REVIEW

File: sortinghat/core/migrations/0009_auto_20210622_1033.py [hunk 1]
```diff
@@ -0,0 +1,37 @@
+# Generated by Django 2.2 on 2021-06-22 15:33
```
Comment (sduenas) [inline sortinghat/core/migrations/0009_auto_20210622_1033.py]: We started using Django 3.1. Please squash the commit you have later with this one.

File: sortinghat/core/models.py [hunk 2]
```diff
@@ -143,6 +145,18 @@ def __str__(self):
         return self.name
 
 
+class Group(MP_Node, EntityBase):
```
Comment (sduenas) [inline sortinghat/core/models.py:148]: Thought we were going to name this class `OrganizationGroup` or something similar. Although if it's no mandatory to assing it to an organization, probably it doesn't make sense. Then, we should look for a different name for the table.

File: sortinghat/core/models.py [hunk 3]
```diff
@@ -143,6 +145,18 @@ def __str__(self):
...
+    name = CharField(max_length=MAX_SIZE_CHAR_INDEX)
+    organization = ForeignKey(Organization, related_name='groups', on_delete=CASCADE, blank=True, null=True)
```
Comment (sduenas) [inline sortinghat/core/models.py:150]: It would be better to split this line in:

```
organization = ForeignKey(Organization, related_name='groups', on_delete=CASCADE,
                          blank=True, null=True)
```

File: tests/test_model.py [hunk 4]
```diff
@@ -106,6 +107,61 @@ def test_last_modified(self):
...
+          org = Organization.objects.create(name='Example')
+          group = Group.add_root(name='subGroup1', organization=org)
+          group.add_child(name='subGroup1', organization=org)
+
+
+    def test_null_organizations(self):
+        """Check if groups can be created without organizations"""
+
+        group = Group.add_root(name='subGroup1')
+        self.assertIsInstance(group,Group)
```
Comment (sduenas) [inline tests/test_model.py:126]: Have you tried to create a second instance with the same name and the organization as null? I think it will allow you when that shouldn't be the case.

File: sortinghat/core/db.py [hunk 5]
```diff
@@ -290,6 +291,77 @@ def delete_organization(trxl, organization):
                        target=op_args['organization'])
 
 
+def add_group(trxl, organization, parent, group_name):
```
Comment (sduenas) [inline sortinghat/core/db.py:294]: As `organization` and `parent` can be `None`, I think it's better to reformulate the definition of the function this way:
```
def add_group(trxl, group_name, organization=None, parent=None):
```

File: sortinghat/core/db.py [hunk 6]
```diff
@@ -290,6 +291,77 @@ def delete_organization(trxl, organization):
...
+    except django.db.utils.IntegrityError as exc:
+        _handle_integrity_error(Group, exc)
+
+    trxl.log_operation(op_type=Operation.OpType.ADD, entity_type='group',
+                       timestamp=datetime_utcnow(), args=op_args,
+                       target=op_args['group_name'])
+    return group
+
+
+def find_group(organization, group_name):
```
Comment (sduenas) [inline sortinghat/core/db.py:338]: Same as before. It should be:

```
def find_group(group_name, organization=None):
```

File: sortinghat/core/db.py [hunk 7]
```diff
@@ -290,6 +291,77 @@ def delete_organization(trxl, organization):
...
+    """Finds a group in an organization, if organization is passed. Otherwise,
+    searches for group with name=`group_name` and organization=`null`.
+
+    Finds a group by its name and organization in the database. When the
+    group does not exist, the function will raise a `NotFoundError`.
```
Comment (sduenas) [inline sortinghat/core/db.py:343]: This also should be:

""Find a group in an organization.

Finds a group by its name in the database by its name. If organization is passed, it looks for a group in that organization. When the group does not exist, the function will raise a `NotFoundError`.

File: sortinghat/core/db.py (hunk 6 above)
Comment (sduenas) [reply sortinghat/core/db.py:338]: This function is should go also where the other `find_*` functions are.

File: tests/test_db.py [hunk 8]
```diff
@@ -203,6 +209,59 @@ def test_organization_not_found(self):
...
+
+        self.assertIsInstance(group, Group)
+        self.assertEqual(group.organization.name, 'Example')
+        self.assertEqual(group.name, 'Example Subgroup')
+        self.assertEqual(group.get_parent(), None)
+
+    def test_group_not_found(self):
+        """Check if an error is raise if group is not found"""
+        with self.assertRaisesRegex(NotFoundError, GROUP_NOT_FOUND_ERROR):
+            db.find_group(self.org, 'subGroup')
```
Comment (sduenas) [inline tests/test_db.py:232]: Leave a blank line between the docstring and the code.

File: tests/test_db.py [hunk 9]
```diff
@@ -203,6 +209,59 @@ def test_organization_not_found(self):
...
+
+    def test_org_name_none(self):
+        """Check if groups linked to no organizations can be found"""
```
Comment (sduenas) [inline tests/test_db.py:235]: Same as before.

File: tests/test_db.py [hunk 10]
```diff
@@ -578,6 +637,198 @@ def test_operations(self):
...
+
+    def test_add_subgroup(self):
+        """Check if a new subgroup is added for specified group"""
+
+        group_name = 'subgroup'
+        group = db.add_group(self.trxl, self.org, None, group_name)
+        subgroup = db.add_group(self.trxl, self.org, group, 'childgroup')
+        self.assertIsInstance(subgroup, Group)
+        self.assertEqual(subgroup.name, 'childgroup')
+        self.assertEqual(subgroup.organization, self.org)
```
Comment (sduenas) [inline tests/test_db.py:670]: Check also that the parent is `subgroup`.

File: tests/test_db.py [hunk 11]
```diff
@@ -578,6 +637,198 @@ def test_operations(self):
...
+        """Load initial dataset"""
+
+        self.user = get_user_model().objects.create(username='test')
+        self.ctx = SortingHatContext(self.user)
+
+        self.trxl = TransactionsLog.open('delete_group', self.ctx)
+        self.org = Organization.objects.create(name='Example')
+
+    def test_delete_group(self):
+        """Check whether it deletes a group"""
```
Comment (sduenas) [inline tests/test_db.py:791]: Leave a blank line between docstring and code.

File: sortinghat/core/api.py [hunk 12]
```diff
@@ -565,6 +568,73 @@ def add_domain(ctx, organization, domain_name, is_top_domain=True):
     return domain
 
 
+@django.db.transaction.atomic
+def add_group(ctx, organization, parent_name, group_name):
```
Comment (sduenas) [inline sortinghat/core/api.py:572]: Because `organization` and `parent_name` can be None, reorganize the definition as explained before.

File: sortinghat/core/api.py [hunk 13]
```diff
@@ -644,6 +714,56 @@ def delete_domain(ctx, domain_name):
     return domain
 
 
+@django.db.transaction.atomic
+def delete_group(ctx, organization, group_name):
```
Comment (sduenas) [inline sortinghat/core/api.py:718]: Same comment with the definition of the function as before.

File: sortinghat/core/api.py [hunk 14]
```diff
@@ -644,6 +714,56 @@ def delete_domain(ctx, domain_name):
...
+    """Remove a group from the registry.
+
+    This function removes the given group of an organization from the registry.
+    Deleting a group also deletes all its subgroups in the organization.
+    Both Organization and Group must exist in the registry. Otherwise, function
+    will raise a `NotFoundError` exception.
```
Comment (sduenas) [inline sortinghat/core/api.py:724]: You should reword this because the group doesn't need to belong to an organization.

File: tests/test_api.py [hunk 15]
```diff
@@ -2026,6 +2029,155 @@ def test_operations(self):
...
+
+    def test_organization_is_none(self):
+        """Check if it fails when organization name is `None`"""
+
+        api.add_group(self.ctx, "Example", None, "suborg")
+        trx_date = datetime_utcnow()
+        with self.assertRaisesRegex(InvalidValueError, GROUP_ORG_NAME_MISSING):
+            api.delete_group(self.ctx, None, "suborg")
+        transactions = Transaction.objects.filter(created_at__gt=trx_date)
+        self.assertEqual(len(transactions), 0)
```
Comment (sduenas) [inline tests/test_api.py:2133]: Please, organize this test (and the other ones) so we can read them better. For example:

```
api.add_group(self.ctx, "Example", None, "suborg")

trx_date = datetime_utcnow()

with self.assertRaisesRegex(InvalidValueError, GROUP_ORG_NAME_MISSING):
     api.delete_group(self.ctx, None, "suborg")

transactions = Transaction.objects.filter(created_at__gt=trx_date)
self.assertEqual(len(transactions), 0)
```

**Review by sduenas — CHANGES_REQUESTED** [review]
In general it looks good but some changes are needed. Please also check and modify the style in the tests and with the docstrings.

File: sortinghat/core/models.py (hunk 2 above)
Comment (Rashmi-K-A) [reply sortinghat/core/models.py:148]: What do you suggest? OrganizationGroup doesnt make much sense since we discussed that some group don't have to be linked to organizations.

File: sortinghat/core/models.py (hunk 3 above)
Comment (Rashmi-K-A) [reply sortinghat/core/models.py:150]: Got it, I forgot to format the file and break down lines with >80 chars, will do

File: sortinghat/core/models.py (hunk 2 above)
Comment (sduenas) [reply sortinghat/core/models.py:148]: To be honest, I'm more towards `Team` to be honest. Many people know about GitHub teams so it's easy to understand what it is.

Comment (Rashmi-K-A) [reply sortinghat/core/models.py:148]: Works for me! I had an initial impression that Team might only feel like small groups (I don't know much about Github teams). However, if you think it will work, we can make the model into Team and call the table 'teams'

File: tests/test_model.py (hunk 4 above)
Comment (Rashmi-K-A) [reply tests/test_model.py:126]: yes, just caught this. I should have put in a unit test for this one. I think I can override the models validate_unique or clean function to catch this.  What are your thoughts?

Comment (Rashmi-K-A) [reply tests/test_model.py:126]: @sduenas I have added a `validate_unique` function to the model and called `group.full_clean()` before saving the objects in `add_group` function in `db.py`. It catches this error. Do you think that will suffice? I will add sufficient tests as well.

File: tests/test_api.py (hunk 15 above)
Comment (Rashmi-K-A) [reply tests/test_api.py]: Done 👍

File: sortinghat/core/api.py (hunk 14 above)
Comment (Rashmi-K-A) [reply sortinghat/core/api.py]: Done 👍

File: sortinghat/core/api.py (hunk 13 above)
Comment (Rashmi-K-A) [reply sortinghat/core/api.py]: Done 👍

File: sortinghat/core/api.py (hunk 12 above)
Comment (Rashmi-K-A) [reply sortinghat/core/api.py]: Done 👍

File: tests/test_db.py (hunk 11 above)
Comment (Rashmi-K-A) [reply tests/test_db.py:796]: Done 👍

File: sortinghat/core/db.py (hunk 6 above)
Comment (Rashmi-K-A) [reply sortinghat/core/db.py]: Done 👍

File: sortinghat/core/db.py (hunk 7 above)
Comment (Rashmi-K-A) [reply sortinghat/core/db.py]: Done 👍

File: tests/test_db.py (hunk 8 above)
Comment (Rashmi-K-A) [reply tests/test_db.py]: Done 👍

File: tests/test_db.py (hunk 9 above)
Comment (Rashmi-K-A) [reply tests/test_db.py:236]: Done 👍

File: tests/test_db.py (hunk 10 above)
Comment (Rashmi-K-A) [reply tests/test_db.py:675]: Done 👍

File: sortinghat/core/db.py (hunk 5 above)
Comment (Rashmi-K-A) [reply sortinghat/core/db.py]: Done 👍

File: sortinghat/core/models.py (hunk 3 above)
Comment (Rashmi-K-A) [reply sortinghat/core/models.py]: Done 👍

File: tests/test_model.py (hunk 4 above)
Comment (sduenas) [reply tests/test_model.py:126]: We need a test for that case to check that it doesn't let it create two groups/teams with the same name and org = NULL.

Comment (Rashmi-K-A) [reply tests/test_model.py]: Done :+1

File: sortinghat/core/db.py [hunk 16]
```diff
@@ -822,13 +924,13 @@ def _handle_integrity_error(model, exc):
...
     m = re.match(_MYSQL_DUPLICATE_ENTRY_ERROR_REGEX,
                  exc.__cause__.args[1])
     if not m:
         raise exc
 
     entity = model.__name__
     eid = m.group('value')
-
+    if model == Team:
+        eid = m.group('value').split('-')[0]
```
Comment (sduenas) [inline sortinghat/core/db.py:935]:
```suggestion

    if model == Team:
        eid = m.group('value').split('-')[0]
        
```

File: sortinghat/core/models.py [hunk 17]
```diff
@@ -143,6 +147,25 @@ def __str__(self):
...
+
+    class Meta:
+        db_table = 'teams'
+        unique_together = ('name', 'organization',)
+
+    def validate_unique(self, exclude=None):
+        if not self.organization:
+            objects = Team.objects.all().filter(name=self.name, organization=self.organization)
+            if len(objects) > 0:
+                raise ValidationError(f"Team {self.name} already exists in the registry")
```
Comment (sduenas) [inline sortinghat/core/models.py:163]: This function is not really supposed to do that. Do you think we can find something else? This method will slow the process. Usually it's better to try to write and get the error that ask first.

File: tests/test_api.py [hunk 18]
```diff
@@ -2026,6 +2029,177 @@ def test_operations(self):
...
+        trx_date = datetime_utcnow()
+
+        with self.assertRaisesRegex(NotFoundError, ORGANIZATION_NOT_FOUND_ERROR.format(name='Exampe')):
+              api.add_team(self.ctx, "subteam", "Exampe", None, )
+
+        transactions = Transaction.objects.filter(created_at__gt=trx_date)
+        self.assertEqual(len(transactions), 0)
+
+    def test_find_parent(self):
+        """ Check if parent is able to be found"""
```
Comment (sduenas) [inline tests/test_api.py:2094]:
```suggestion
        """ Check if parent is able to be found"""
        
```

File: tests/test_db.py [hunk 19]
```diff
@@ -578,6 +640,226 @@ def test_operations(self):
...
+
+        team1 = db.add_team(self.trxl, "team1", self.org, None)
+        team2 = db.add_team(self.trxl, "team2", self.org, None)
+
+        teams = Team.objects.all()
+
+        self.assertIsInstance(team1, Team)
+        self.assertIsInstance(team2, Team)
+        self.assertEqual(len(teams), 2)
+
```
Comment (sduenas) [inline tests/test_db.py:779]: Remove this line here.

**Review by sduenas — CHANGES_REQUESTED** [review]
In overall it looks ok but I don't like much the way you're solving the integrity problem with the same name team without organization.
You also need to remove that merge commit that is there.

File: sortinghat/core/models.py (hunk 17 above)
Comment (Rashmi-K-A) [reply sortinghat/core/models.py:163]: Yes, we need to figure out some way to add this validation though. Because if teams have orgs as NULL, the inbuilt validation just doesn't work. This is going to test for every node that has orgs as NULL and I agree that its going to affect the performance if we have too many teams without orgs.  Is there any alternative that you can suggest or shall I dig around more and see what I can find?

**Comment by Rashmi-K-A** [comment]
> You also need to remove that merge commit that is there.
Yes, I have a call with Venu today to fix that. I thought its better to let him show me how to fix that the first time!

File: sortinghat/core/models.py (hunk 17 above)
Comment (sduenas) [reply sortinghat/core/models.py:163]: I don't exactly remember how we do this with the identities. The problem should be the same. Can you check it? Because I couldn't find it.

**Comment by Rashmi-K-A** [comment]
@sduenas  I have fixed the commits (thanks to @vchrombie!). We were wondering about the changelog entry. Should that be added as part of this PR?

**Comment by sduenas** [comment]
> @sduenas I have fixed the commits (thanks to @vchrombie!). We were wondering about the changelog entry. Should that be added as part of this PR?

Maybe we can skip that part until the implementation with graphql + ui is done.

**Comment by sduenas** [comment]
> @sduenas I have fixed the commits (thanks to @vchrombie!). We were wondering about the changelog entry. Should that be added as part of this PR?

@Rashmi-K-A would you mind to change the style of the commits again? The body of the message should be limited to 70 characters or so. Now they are very long and only good for large screens.

**Comment by vchrombie** [comment]
@Rashmi-K-A 
Good work with the re-writing the commits, but I think we have some more things to do.
You can truncate the commit description into multiple lines.

**Comment by vchrombie** [comment]
> @Rashmi-K-A would you mind to change the style of the commits again? The body of the message should be limited to 70 characters or so. Now they are very long and only good for large screens.

Jinx! 😄 🥂

**Comment by vchrombie** [comment]
@Rashmi-K-A let me know if you need help with anything.

**Comment by Rashmi-K-A** [comment]
Thanks @vchrombie @sduenas !! I have fixed it now 💯  Let me know if I need to change anything else!

**Review by sduenas — APPROVED** [review]
LGTM

**Comment by sduenas** [comment]
Merged. Thanks for your contribution!
